### PR TITLE
Clean up JNI and improve freeing native memory (CBL-130)

### PR DESCRIPTION
**Pull request by pasin**
1. Used only public API except for RefCounted
    1. Use FLSlice instead of slice.
    2. Use FLSliceResult instead of AllocSlice.
    3. Use FLEncoder instead of Encoder and fix some bugs in FLEncoder.
2. Instead of freeing FLSliceResult in finalize(), make sure the consumer codes call free() method. This will match the way FLSliceResult is used natively.
3. Instead of freeing FLEncoder in finalize(), make sure the customer codes call free() method. This will match the way FLSliceResult is used natively.
4. Ensured to reset the shared FLEncoder after using it.
5. Fixed bug in jstringSlice C++ class when the string is null. This seems to cause a lot of memory issue (bad_alloc) when changing to use FLSlice instead of slice.

### CODE REVIEW

**Comment by pasin** [comment]
I fixed error reported by checkstyle and findbugs and did force-push.

File: src/main/cpp/native_glue.hh [hunk 1]
```diff
@@ -75,17 +71,15 @@ namespace litecore {
...
-
-            operator slice() { return _slice; }
+                      _critical(s._critical) { s._slice = kFLSliceNull; }
 
-            operator C4Slice() { return {_slice.buf, _slice.size}; }
+            operator FLSlice() { return _slice; }
 
             // Copies a Java byte[] to an alloc_slice
-            static alloc_slice copy(JNIEnv *env, jbyteArray jbytes);
+            static FLSliceResult copy(JNIEnv *env, jbyteArray jbytes);
```
Comment (jayahariv) [inline src/main/cpp/native_glue.hh]: // Copies a Java byte[] to an " `FLSliceResult`  "

File: src/main/cpp/native_c4query.cc [hunk 2]
```diff
@@ -128,9 +128,9 @@ Java_com_couchbase_lite_internal_core_C4Query_createIndex(JNIEnv *env, jclass cl
     jstringSlice expressionsJSON(env, jexpressionsJSON);
     jstringSlice language(env, jlanguage);
     C4IndexOptions options = {};
-    slice sLang = language;
+    FLSlice sLang = language;
     if(sLang.buf != NULL)
-        options.language = (const char *)sLang.buf;
+        options.language = language.c_str();
```
Comment (jayahariv) [inline src/main/cpp/native_c4query.cc]: do we need the `if(sLang.buf != NULL)`? options.language == NULL

File: src/main/cpp/native_glue.cc [hunk 3]
```diff
@@ -157,16 +159,17 @@ namespace litecore {
                   _jbytes(jbytes),
                   _critical(critical) {
             if (jbytes == nullptr) {
-                _slice = nullslice;
+                _slice = kFLSliceNull;
                 return;
             }
 
             jboolean isCopy;
```
Comment (jayahariv) [inline src/main/cpp/native_glue.cc]: what are we doing with this `jboolean isCopy;`? Seems like it is unused

Comment (pasin) [reply src/main/cpp/native_glue.cc]: I don't think it's necessary, but it's out of scope of the purpose of this PR.

File: src/main/cpp/native_c4query.cc (hunk 2 above)
Comment (pasin) [reply src/main/cpp/native_c4query.cc]: I don't think we need it. I will remove it and test.

File: src/main/cpp/native_glue.hh (hunk 1 above)
Comment (pasin) [reply src/main/cpp/native_glue.hh]: Will fix this. Thanks.

File: src/main/cpp/native_glue.cc (hunk 3 above)
Comment (pasin) [reply src/main/cpp/native_glue.cc]: I will remove it. Thanks.

**Pull request merged**